### DISCUSSION TITLE
Parsing strings to chords

### DISCUSSION THREAD

**Discussion by mainrs** [ozankasikci/fantasy-in-rust#22] [General] [answered]
Hey! I have a simple use-case where I want to parse strings into the correct chords:
```
D#maj7
C# (implies major)
C#min/m/minor
```
Can this be done with this crate?

**Comment by henryksloan** [comment, marked as answer]
Hi! The library provides `from_regex` functions for most domains, including chords and scales. The general form of the regex is `<root> <quality> <number>`, where quality and number are optional. That means that `C#` will work as you indicate. To get D# major 7, you would have to pass `D# major seventh` or `D# maj seventh`, as the number must be fully spelled out (though number and quality are case insensitive, and number can be omitted). C# minor can be retrieved by passing `C# minor` or `C# min`.

It wouldn't be a huge modification to allow abbreviations. In fact, the regex parses are already case insensitive, and effectively ignore whitespace. One would just have to extend the regexes in `number.rs` and `quality.rs` in such a way that accounts for overlaps. For example, the regex for `maj(or)?\s*(seventh | 7)` should come before `seventh | 7` (meaning dominant)`.

**Reply by mainrs** [reply]
Good to hear that this is indeed possible. I can open a PR if you're fine with regex extensions.

**Reply by henryksloan** [reply]
I think that would be great, though it'll be up to @ozankasikci's approval. But I think everyone would be in favor of more permissive regexes! If you get a chance, the tests of number and quality regex parsing (`test_number.rs` and `test_quality.rs` respectively) could use some more tests, so feel free to add tests relevant to your changes.

**Reply by ozankasikci** [reply]
@henryksloan thanks for the detailed reply!

This is indeed possible and would be a nice improvement for the regex parser. Feel free to work on that @SirWindfield if you want 👍

**Reply by XBagon** [reply]
By the way, I'm planning to split up the parsing in the CLI with my `structopt` rewrite. `PitchClass`, `Mode`, etc. are ready independently and put together afterwards, that would make for a more strict, straightforward API, while also not having to rely on regex too heavily. The `from_regex` functions of `Chord` and `Scale` would still be useful for library purposes probably, and I could also add a `--regex` flag to the CLI.
